src/lang.rs

```diff
@@ -884,6 +908,80 @@ impl Context for CallStack {
         }
     }
 
+    fn get_mutable(&mut self, name: String) -> EvalResult {
+        let mut env = self.env();
+        // whether the object was found immediately
+        let mut caller_env = true;
+        loop {
+            let value = if caller_env {
+                // search in this environment for value by name
+                let Some(value) = env
+                    .values
+                    .borrow_mut()
+                    .get(&name)
+                    .map(|o| (*o).mutable_view())
+                else {
+                    caller_env = false;
+                    // if not found, search through parent if available
+                    if let Some(parent) = &env.parent {
+                        env = parent.clone();
+                        continue;
+                    } else {
+                        break;
+                    }
+                };
+                value
+            } else {
+                let Some(value) = env.values.borrow_mut().get(&name).map(|o| (*o).lazy_copy())
+                else {
+                    // if not found, search through parent if available
+                    if let Some(parent) = &env.parent {
+                        env = parent.clone();
+                        continue;
+                    } else {
+                        break;
+                    }
+                };
+                value
+            };
```

Comment:
Maybe we can break out the environment parent traversal into a separate helper, such as `Context::find_env_for` that returns the first environment walking up the search path that contains a symbol. 

In pseudo-code

```rust
env_containing_name = env.find_env_containing(name)
if env_containing_name == env {
  // env.get..borrow_mut..mutable_view
} else {
  // env_containing_name.get..borrow_mut..lazy_copy
}
```

Let me know if I'm missing something - it's very possible that I'm overlooking some key complexity